File: src/features/provedores/clientes.py
from collections.abc import Mapping
from typing import Any, cast

from src.shared.http_readonly import ReadOnlyHttpClient
from src.shared.runtime import Settings

_PAGINATION_KEYS = {
    "totalRegistros": "total_items",
    "totalPaginas": "total_pages",
    "numeroPagina": "page",
    "tamanhoPagina": "page_size",
}


def _normalize_page(payload: Any, *item_keys: str) -> dict[str, Any]:
    if isinstance(payload, list):
        return {"items": payload, "pagination": {}}

    if isinstance(payload, Mapping):
        page = cast(Mapping[str, Any], payload)
        items: Any = page
        for item_key in item_keys:
            if item_key in page:
                items = page[item_key]
                break
        pagination = {
            normalized_key: page[upstream_key]
            for upstream_key, normalized_key in _PAGINATION_KEYS.items()
            if upstream_key in page
        }
        return {"items": items, "pagination": pagination}

    return {"items": payload, "pagination": {}}


class ComprasClient:
    base_url = "https://dadosabertos.compras.gov.br"

    @staticmethod
    def normalize_page(payload: Any) -> dict[str, Any]:
        return _normalize_page(payload, "resultado", "data")

    def client(self, *, max_document_bytes: int | None = None) -> ReadOnlyHttpClient:
        if max_document_bytes is not None and max_document_bytes <= 0:
            raise ValueError("max_document_bytes must be positive")
        settings = Settings()  # pyright: ignore[reportCallIssue]
        return ReadOnlyHttpClient(
            self.base_url,
            requests_per_second=settings.http_requests_per_second,
            max_concurrency=settings.compras_max_concurrency,
            provider="compras",
            max_retries=settings.http_max_retries,
            timeout=settings.http_timeout,
            max_document_bytes=(
                settings.max_document_bytes
                if max_document_bytes is None
                else max_document_bytes
            ),
        )


class PncpClient:
    base_url = "https://pncp.gov.br/api/pncp"

    @staticmethod
    def normalize_page(payload: Any) -> dict[str, Any]:
        return _normalize_page(payload, "data")

    def client(self, *, max_document_bytes: int | None = None) -> ReadOnlyHttpClient:
        if max_document_bytes is not None and max_document_bytes <= 0:
            raise ValueError("max_document_bytes must be positive")
        settings = Settings()  # pyright: ignore[reportCallIssue]
        return ReadOnlyHttpClient(
            self.base_url,
            requests_per_second=settings.http_requests_per_second,
            max_concurrency=settings.pncp_max_concurrency,
            provider="pncp",
            max_retries=settings.http_max_retries,
            timeout=settings.http_timeout,
            max_document_bytes=(
                settings.max_document_bytes
                if max_document_bytes is None
                else max_document_bytes
            ),
        )

File: src/features/provedores/clientes.ts
import { ReadOnlyHttpClient } from "../../shared/http_readonly.js";
import { loadSettings } from "../../shared/runtime.js";

const PAGINATION_KEYS: ReadonlyMap<string, string> = new Map([
  ["totalRegistros", "total_items"],
  ["totalPaginas", "total_pages"],
  ["numeroPagina", "page"],
  ["tamanhoPagina", "page_size"],
]);

export interface NormalizedPage {
  items: unknown;
  pagination: Record<string, unknown>;
}

function normalizePage(payload: unknown, ...itemKeys: string[]): NormalizedPage {
  if (Array.isArray(payload)) return { items: payload, pagination: {} };

  if (typeof payload === "object" && payload !== null) {
    const page = payload as Record<string, unknown>;
    let items: unknown = page;
    for (const itemKey of itemKeys) {
      if (Object.hasOwn(page, itemKey)) {
        items = page[itemKey];
        break;
      }
    }
    const pagination: Record<string, unknown> = {};
    for (const [upstreamKey, normalizedKey] of PAGINATION_KEYS) {
      if (Object.hasOwn(page, upstreamKey)) {
        pagination[normalizedKey] = page[upstreamKey];
      }
    }
    return { items, pagination };
  }

  return { items: payload, pagination: {} };
}

interface ClientOptions {
  maxDocumentBytes?: number;
}

function buildClient(
  baseUrl: string,
  provider: "compras" | "pncp",
  maxConcurrency: number,
  settings: ReturnType<typeof loadSettings>,
  opts: ClientOptions | undefined,
): ReadOnlyHttpClient {
  return new ReadOnlyHttpClient(baseUrl, {
    requestsPerSecond: settings.httpRequestsPerSecond,
    maxConcurrency,
    provider,
    maxRetries: settings.httpMaxRetries,
    timeout: settings.httpTimeout,
    maxDocumentBytes: opts?.maxDocumentBytes ?? settings.maxDocumentBytes,
  });
}

export class ComprasClient {
  static readonly baseUrl = "https://dadosabertos.compras.gov.br";

  static normalizePage(payload: unknown): NormalizedPage {
    return normalizePage(payload, "resultado", "data");
  }

  client(opts?: { maxDocumentBytes?: number }): ReadOnlyHttpClient {
    if (opts?.maxDocumentBytes !== undefined && opts.maxDocumentBytes <= 0) {
      throw new Error("max_document_bytes must be positive");
    }
    const settings = loadSettings();
    return buildClient(
      ComprasClient.baseUrl,
      "compras",
      settings.comprasMaxConcurrency,
      settings,
      opts,
    );
  }
}

export class PncpClient {
  static readonly baseUrl = "https://pncp.gov.br/api/pncp";

  static normalizePage(payload: unknown): NormalizedPage {
    return normalizePage(payload, "data");
  }

  client(opts?: { maxDocumentBytes?: number }): ReadOnlyHttpClient {
    if (opts?.maxDocumentBytes !== undefined && opts.maxDocumentBytes <= 0) {
      throw new Error("max_document_bytes must be positive");
    }
    const settings = loadSettings();
    return buildClient(
      PncpClient.baseUrl,
      "pncp",
      settings.pncpMaxConcurrency,
      settings,
      opts,
    );
  }
}
